akka-docs/src/test/java/jdocs/stream/operators/sourceorflow/MapAsyncs.java

```diff
@@ -99,20 +103,153 @@ private void runUnordered() {
+
+                    if (random.nextBoolean() || countsPerEntity.isEmpty()) {
+                      entityId = random.nextInt(n);
+                    } else {
+                      int m = random.nextInt(countsPerEntity.size()) + 1;
+                      Iterator<Integer> keysIterator = countsPerEntity.keySet().iterator();
+
+                      Integer selected = Integer.valueOf(-1);
+                      int i = 0;
+                      while (keysIterator.hasNext() && i < m) {
+                        if (i == m - 1) {
+                          selected = keysIterator.next();
+                        } else {
+                          keysIterator.next();
+                        }
+                        i++;
+                      }
+                      entityId = selected.intValue();
+                    }
+
+                    if (entityId < 0) {
+                      throw new AssertionError("entityId should be non-negative, was " + entityId);
+                    }
+
+                    int seqNr =
+                        countsPerEntity
+                                .getOrDefault(Integer.valueOf(entityId), Integer.valueOf(0))
+                                .intValue()
+                            + 1;
+                    countsPerEntity.put(Integer.valueOf(entityId), Integer.valueOf(seqNr));
+
+                    return Collections.singletonList(new EntityEvent(entityId, seqNr));
+                  };
+                })
+            .take(1000);
+
+    // #mapAsyncPartitioned
+    Function<EntityEvent, Integer> partitioner =
+        (event) -> {
+          Integer partition = Integer.valueOf(event.entityId);
+          System.out.println("Assigned event " + event + " to partition " + event.entityId);
+          return partition;
+        };
+
+    eventsForEntities
+        .statefulMapConcat(
+            () -> {
+              final int[] offset = new int[1]; // trick to close over an int...
+              offset[0] = 0;
+
+              return (event) -> {
+                System.out.println(
+                    "Received event " + event + " at offset " + offset[0] + " from message broker");
+                offset[0]++;
+                return Collections.singletonList(event);
+              };
+            })
+        .mapAsyncPartitioned(
+            10,  // parallelism
+            1,   // perPartition
+            partitioner,
+            (event, partition) -> {
+              System.out.println("Processing event " + event + " from partition " + partition);
+
+              CompletionStage<String> cs;
+              CompletableFuture<String> cf =
+                  CompletableFuture.completedFuture(
+                      partition.toString() + "-" + event.sequenceNumber());
+              if (random.nextBoolean()) {
+                cs = Patterns.after(Duration.ofMillis(random.nextInt(1000)), system, () -> cf);
+              } else {
+                cs = cf;
+              }
```

Comment:
Let's hide the internal logic to create a CS with some random completion here into a method not included in the sample snippet. Ideally I'd actually like to see an ask/sharding entity ref ask here, since that is the primary use case, that'd be hard keeping it runnable but I don't think that really is a priority as that anyway requires users to go to sources/check out Akka codebase etc.

Reply:
Hmm... while an ask to a sharded entity is definitely the motivation, streams definitely seems a lower-level, more core layer than cluster sharding and there seems to be a general tendency to not have further-in layers refer to further-out layers (if that makes sense).

If @octonato's #30934 were available, that would solve this (by doing an ask to an ActorSystem-level entity), but perhaps in the interim a little sleight-of-hand in naming things for the docs is called for.

Reply:
Yeah, just a method returning CompletionStage is fine. I think the current pretend-doing-work just makes the example harder to follow (because users may think it is something they will need to do themselves) so hiding that was the most important part.

In other places we have done nasty tricks along the lines of making a local API/impl that _looks_ like the entity ref lookup returning something that looks like entity ref and can be asked. (I think that was for typed ActorContext but didn't find it with a quick search)